### PR TITLE
Added the eucan-connect code

### PR DESCRIPTION
**Pull request by dtroelofsprins**
#### Checklist
- [x] Functionality works & meets specifications
- [x] Code reviewed
- [x] Code unit/system tested
- [x] User documentation updated
- [x] Clean commits
- [x] Updated CHANGELOG.md

### CODE REVIEW

**Review by tommydeboer — CHANGES_REQUESTED** [review]

File: src/molgenis/eucan_connect/mica.py [hunk 1]
```diff
@@ -0,0 +1,761 @@
...
+        retrieving a list with study IDs. Depending on the value of param amount
+        only the total amount of studies or the real study information is returned
+        @param start: study number (row) to start from, default is 0
+        @param batch_size: number of studies to be returned
+                           if != totalCount, a loop should be created
+        """
+        studies = []
+        study_url = self.catalogue.catalogue_url + "/ws/studies/_rql"
+        study_query = (
+            "query=study(limit(" + str(start) + "," + str(batch_size) + "),fields(*))"
```
Comment (tommydeboer) [inline src/molgenis/eucan_connect/mica.py:112]: You can use string formatting here, which reads a bit nicer:

```suggestion
            f"query=study(limit({str(start)},{str(batch_size)}"),fields(*))"
```

**Review by tommydeboer — CHANGES_REQUESTED** [review]

File: src/molgenis/eucan_connect/errors.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import requests
+
+from molgenis.eucan_connect.model import Catalogue
+
+
+@dataclass(frozen=True)
+class EucanWarning:
+    """
+    Class that contains a warning message. Use this when a problem occurs that
+    shouldn't cancel the current action (for example staging or publishing).
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/errors.py]: Changed

File: src/molgenis/eucan_connect/errors.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    Raise this exception when an error occurs that we can not recover from.
+    """
+
+    pass
+
+
+@dataclass
+class ErrorReport:
+    """
+    Summary object. Stores errors and warnings that occurred for each node.
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/errors.py]: Changed

File: src/molgenis/eucan_connect/eucan.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        self, catalogue: Catalogue, report: ErrorReport, importer: Importer
+    ):
+        # Get the data from the source catalogue(s)
+        if catalogue.catalogue_type == "BirthCohorts":
+            # Get the data from the source catalogue type birth cohorts
+            print("Birth cohort data. No module available yet!")
+            # catalogue_data = self._get_bc_data(catalogue)
+        if catalogue.catalogue_type == "Mica":
+            # Get the data from the source catalogue type Mica
+            catalogue_data = self._get_mica_data(catalogue)
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/eucan.py:69]: Your right!

File: src/molgenis/eucan_connect/eucan.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        return report
+
+    @requests_error_handler
+    def _import_catalogue(
+        self, catalogue: Catalogue, report: ErrorReport, importer: Importer
+    ):
+        # Get the data from the source catalogue(s)
+        if catalogue.catalogue_type == "BirthCohorts":
+            # Get the data from the source catalogue type birth cohorts
+            print("Birth cohort data. No module available yet!")
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/eucan.py]: It actually could be removed at all, although this catalogue type is available in the database. I left it there and in stead of a print or warning raise an EucanError.

File: src/molgenis/eucan_connect/eucan.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        self._import_catalogue_data(catalogue_data, importer, report)
+
+    @requests_error_handler
+    def _get_mica_data(self, catalogue: Catalogue):
+        try:
+            self.printer.print_sub_header(
+                f"📥 Get data of source catalogue {catalogue.description}"
+            )
+            return Mica(
+                self.session, IsoCountryData, self.ref_data, self.printer
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/eucan.py]: Changed this, a bit strange how this ended up in the final version.

File: src/molgenis/eucan_connect/eucan.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+                f"Error retrieving data of catalogue {catalogue.description}"
+            ) from e
+
+    def _add_new_ref_data(
+        self, ref_data: RefData, importer: Importer, report: ErrorReport
+    ):
+        """
+        Inserts new reference data into the EUCAN-Connect Catalogue
+        """
+        self.warnings = []
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/eucan.py:99]: True

File: src/molgenis/eucan_connect/eucan.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    def _import_catalogue_data(
+        self, catalogue_data: CatalogueData, importer: Importer, report: ErrorReport
+    ):
+        """
+        Inserts the data of the source catalogue to the EUCAN-Connect Catalogue
+        This happens in two phases:
+        1. All source catalogue data are removed from the EUCAN-Connect Catalogue
+        2. Data from the source catalogue are inserted into EUCAN-Connect Catalogue
+        """
+        self.warnings = []
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/eucan.py:117]: True

File: src/molgenis/eucan_connect/importer.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    """
+    This class is responsible for uploading the data into the EUCAN-Connect Catalogue
+    """
+
+    def __init__(self, session: EucanSession, printer: Printer):
+        self.session = session
+        self.printer = printer
+        self.warnings: List[EucanWarning] = []
+
+    def import_data(self, data: dict, data_cat: str) -> List[EucanWarning]:
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/importer.py:20]: You are right. I've changed this and created indeed two separate functions.

File: src/molgenis/eucan_connect/eucan.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    ):
+        """
+        Inserts new reference data into the EUCAN-Connect Catalogue
+        """
+        self.warnings = []
+        self.printer.print_sub_header("📤 Importing new reference data")
+        self.printer.indent()
+
+        warnings = importer.import_data(ref_data, "RefData")
+        report.add_warnings(ref_data.ref_entity, warnings)
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/eucan.py:103]: I thought it was the correct value, but you are right, this should be the catalogue. I've changed this.

File: src/molgenis/eucan_connect/model.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+# @dataclass(frozen=True)
+# class EucanEntities:
+#     """
+#     Object containing data of the EUCAN-Cannect Catalogue entities.
+#     """
+#
+#     events: List[dict]
+#     populations: List[dict]
+#     persons: List[dict]
+#     studies: List[dict]
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/model.py:236]: Done

File: src/molgenis/eucan_connect/printer.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
+# from molgenis.eucan_connect.errors import EucanError #, EucanWarning, EucanReport
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/printer.py:1]: Done

File: src/molgenis/eucan_connect/printer.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    #             if source in report.warnings:
+    #                 message += f" with {len(report.warnings[node])} warning(s)"
+    #         elif source in report.warnings:
+    #             message = (
+    #                 f"⚠️ Source catalogue {source.Source} finished successfully with "
+    #                 f"{len(report.warnings[node])} warning(s)"
+    #             )
+    #         else:
+    #             message = f"✅ Source catalogue {source.Source} finished successfully"
+    #         self.print(message)
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/printer.py:71]: Yes, this was a todo, now done.

File: src/molgenis/eucan_connect/eucan_client.py [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+class EucanSession(ExtendedSession):
+    """
+    A session with a EUCAN-Connect Catalogue. Contains methods to get source catalogues,
+    their data and EUCAN-Connect data.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(EucanSession, self).__init__(*args, **kwargs)
+        self.printer = Printer()
+        self.warnings: List[EucanWarning] = []
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/eucan_client.py:63]: True, removed them.

File: src/molgenis/eucan_connect/importer.py [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+            add = list()
+            for reference in ref_data.table_by_type[table_type].rows:
+                if reference[id_attr] not in existing_ids:
+                    add.append(reference)
+
+            self.printer.print(f"Importing {len(add)} rows in {entity_type_id}")
+            try:
+                self.session.add_batched(entity_type_id, add)
+            except MolgenisRequestError as e:
+                raise EucanError(f"Error importing rows to {entity_type_id}") from e
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/importer.py:89]: True, but then the data must be modelled in the right way like NodeData/CatalogueData/RefData

File: src/molgenis/eucan_connect/eucan_client.py [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+        :return: a list of dictionaries
+        """
+
+        eucan_ref_data = self.get(
+            entity_type_id, batch_size=10000, attributes="id,label"
+        )
+
+        ref_data = utils.to_upload_format(eucan_ref_data)
+
+        return ref_data
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/eucan_client.py:139]: Changed the return type to List[dict]

File: src/molgenis/eucan_connect/importer.py [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+            self.printer.print(f"Delete existing rows in {table.type.base_id}")
+            try:
+                self._delete_rows(table, catalogue_data.catalogue)
+            except MolgenisRequestError as e:
+                raise EucanError(
+                    f"Error deleting existing rows from {table.type.base_id}"
+                ) from e
+        self.printer.dedent()
+
+        self.printer.indent()
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/importer.py:56]: Changed.

File: src/molgenis/eucan_connect/mica.py (hunk 1 above)
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/mica.py:112]: Implemented

File: src/molgenis/eucan_connect/eucan_client.py [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+        """
+
+        tables = dict()
+        for ref_entity in RefEntity.get_ref_entities():
+            id_ = ref_entity.base_id
+            tables[ref_entity] = RefTable.of(
+                table_type=ref_entity, rows=self._get_ref_entity_data(id_)
+            )
+
+        return RefData.from_dict(ref_entity=RefEntity, tables=tables)
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/eucan_client.py:113]: Removed the refentity=RefEntity, as it turnes out that it has no function. Tested the code (both technical as well as functional) and without this all works OK.

File: src/molgenis/eucan_connect/model.py [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+
+    def get_country_id(self, country_description: str) -> str:
+        iso_country = None
+        for iso_type in self.iso_country_data[0].keys():
+            iso_country = next(
+                (
+                    country
+                    for country in self.iso_country_data
+                    if country[iso_type] == country_description
+                ),
```
Comment (tommydeboer) [inline src/molgenis/eucan_connect/model.py:149]: Didn't know about this one, nice! :) I think this notation is a bit easier to read though:

```python
iso_country = next(
    filter(lambda country: country[iso_type] == country_description, self.iso_country_data),
    False)
```

File: src/molgenis/eucan_connect/model.py [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+        return replacements
+
+    def add_new_ref(self, ref_entity, new_ref, ref_description):
+        refs = self.table_by_type[RefEntity(ref_entity)].rows
+        refs.append({"id": new_ref, "label": ref_description})
+        self.table_by_type[RefEntity(ref_entity)] = RefTable.of(
+            table_type=ref_entity, rows=refs
+        )
+
+    #  @property TODO: is dit een property?
```
Comment (tommydeboer) [inline src/molgenis/eucan_connect/model.py:229]: Resolve this TODO

File: src/molgenis/eucan_connect/model.py [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+    def all_refs(self, ref_entity) -> List[str]:
+        all_refs = list(self.table_by_type[RefEntity(ref_entity)].rows_by_id.keys())
+        return all_refs
+
+    @staticmethod
+    def from_dict(tables: Dict[RefEntity, RefTable]) -> "RefData":
+        return RefData(
+            # ref_entity=ref_entity,
```
Comment (tommydeboer) [inline src/molgenis/eucan_connect/model.py:237]: 🧼 🧽

File: src/molgenis/eucan_connect/model.py [hunk 22]
```diff
@@ -0,0 +1,239 @@
...
+            type=table_type,
+            rows_by_id=rows_by_id,
+        )
+
+
+@dataclass()
+class RefData:
+    """Container object storing the reference entity data."""
+
+    # ref_entity: RefEntity
```
Comment (tommydeboer) [inline src/molgenis/eucan_connect/model.py:207]: 🧼 🧽

File: tests/conftest.py [hunk 23]
```diff
@@ -7,4 +7,288 @@
...
+@pytest.fixture
+def printer() -> MagicMock:
+    return MagicMock()
+
+
+@pytest.fixture
+def eucan(session, printer, ref_data) -> Eucan:
+    eucan = Eucan(session)
+    eucan.printer = printer
+    # eucan.iso_country_data: country_data
```
Comment (tommydeboer) [inline tests/conftest.py:70]: 🧼

File: tests/test_eucan.py [hunk 24]
```diff
@@ -0,0 +1,248 @@
...
+
+    # TODO werkt ook niet, zegt dat ie niet gecalled wordt
+    # importer_init.import_catalogue_data.assert_called_once_with(lc_catalogue_data)
+
+    # TODO dit werkt ook niet, de import_reference_data lijkt niet gecalled te worden
+    # assert importer_init.mock_calls == [
+    #     mock.call(session, eucan.printer),
+    #     mock.call().import_catalogue_data(lc_catalogue_data),
+    #     mock.call().import_reference_data(eucan.ref_data),
+    # ]
```
Comment (tommydeboer) [inline tests/test_eucan.py:152]: Same here

File: tests/test_importer.py [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+        TableType.POPULATIONS: populations,
+        TableType.STUDIES: studies,
+    }
+    catalogue_data = fake_catalogue_data
+    importer._get_eucan_ids = MagicMock()
+
+    importer.import_catalogue_data(catalogue_data)
+
+    # Ik zou verwachten dat deze meerdere keren werd gecalled, maar zegt 0 keer
+    # assert importer._get_eucan_ids.assert_called_once()
```
Comment (tommydeboer) [inline tests/test_importer.py:34]: Same here

File: tests/test_lifecycle.py [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+        converted_ndarrays = converted_df[np_ndarray].apply(
+            lambda x: list(x).sort() if x is not np.nan else x
+        )
+        check_ndarrays = lifecycle_converted_df[np_ndarray].apply(
+            lambda x: list(x).sort() if x is not np.nan else x
+        )
+
+        pd.testing.assert_series_equal(converted_ndarrays, check_ndarrays)
+        # assert np.array_equal(converted_ndarrays, check_ndarrays, equal_nan=True)
+        # np.testing.assert_array_equal(converted_ndarrays, check_ndarrays)
```
Comment (tommydeboer) [inline tests/test_lifecycle.py:79]: 🧼 or enable

**Review by tommydeboer — CHANGES_REQUESTED** [review]

File: README.md [hunk 27]
```diff
@@ -1,84 +1,12 @@
...
-tox
-```
-
->Note: If you want automatic code formatting in your IDE, you need to configure file watchers
-  for `black` (the code formatter) and `isort` (the formatter for import statements). Here
-  are some examples of how to configure them in `PyCharm`:
->
-> ![img.png](.img/example_black_config.png)
-> ![img_1.png](.img/example_isort_config.png)
+These tools can be used as a library in a script or as a command line tool. In both
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply README.md:11]: True, I've removed it.

File: src/molgenis/eucan_connect/mica.py [hunk 28]
```diff
@@ -0,0 +1,761 @@
...
+                    table_type=table_type,
+                    meta=meta,
+                    rows=self._to_populations(table_type, study_data),
+                )
+            elif id_ == "eucan_study":
+                tables[table_type] = Table.of(
+                    table_type=table_type,
+                    meta=meta,
+                    rows=self._to_studies(table_type, study_data),
+                )
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/mica.py]: That's useful to know, this piece of code is rewritten and not used anymore, but I'll check if I've used something like this in another place.

File: src/molgenis/eucan_connect/mica.py [hunk 29]
```diff
@@ -0,0 +1,761 @@
...
+            studies_summary = response.json()["studyResultDto"][
+                "obiba.mica.StudyResultDto.result"
+            ]["summaries"]
+
+        for study in studies_summary:
+            studies.append(study["id"])
+
+        return studies
+
+    def get_mica_study_data(self, mica_studies: List):
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/mica.py]: I didn't know what to use when it was used a "normal" list (for dict's I knew), thanks!

File: src/molgenis/eucan_connect/eucan.py [hunk 30]
```diff
@@ -0,0 +1,294 @@
...
+                # list(map etc) does not work with replace
+                df[col] = df[col].apply(
+                    lambda x: [i.replace(invalid_character, replacement) for i in x]
+                    if x is not np.nan
+                    else x
+                )
+
+        return df
+
+    def _create_catalogue_data(
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/eucan.py:210]: True and done

File: src/molgenis/eucan_connect/eucan.py [hunk 31]
```diff
@@ -0,0 +1,294 @@
...
+                table_type=table_type,
+                meta=meta,
+                rows=self.get_uploadable_data(catalogue, df_in, table),
+            )
+
+        return CatalogueData.from_dict(
+            catalogue=catalogue, source=catalogue.description, tables=tables
+        )
+
+    def get_uploadable_data(
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/eucan.py:237]: I agree, included it into the EucanSession

File: src/molgenis/eucan_connect/importer.py [hunk 32]
```diff
@@ -0,0 +1,134 @@
...
+                if reference[id_attr] not in existing_ids:
+                    add.append(reference)
+
+            if len(add) > 0:
+                self.printer.print(
+                    f"{len(add)} new row(s) will be imported " f"in {entity_type_id}"
+                )
+            else:
+                self.printer.print(
+                    f"No new data needs to be imported " f"in {entity_type_id}"
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/importer.py:82]: 🤔, that's really strange, I think reformatting when committing has done this. I've changed it.

File: src/molgenis/eucan_connect/lifecycle.py [hunk 33]
```diff
@@ -0,0 +1,399 @@
...
+
+        # Extract the contact information, data collection events and populations
+        # As either one or more of them are missing for some cohorts exclude missing
+        # rows when extracting the data
+        df_cohorts = self._extract_data(json_data, df_cohorts, table_prefix)
+
+        # Rename the columns of the extracted data
+        df_cohorts.rename(
+            columns={  # "events_areasOfInformation":
+                #    "events_type_administrative_databases",
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/lifecycle.py:136]: I've checked it again, and decided areasOfInformation can indeed be mapped to type of administrative databases.

File: src/molgenis/eucan_connect/lifecycle.py [hunk 34]
```diff
@@ -0,0 +1,399 @@
...
+        )
+
+        df_converted["temp_contacts"] = [
+            "Yes" if x == "Contact person" else np.nan
+            for x in df_converted["persons_contribution_types"]
+        ]
+        df_converted["temp_contacts"] = df_converted["persons_id"].where(
+            df_converted["temp_contacts"] == "Yes", np.nan
+        )
+        # TODO Discuss: As not all persons have a contribution type,
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/lifecycle.py:243]: Done

File: src/molgenis/eucan_connect/lifecycle.py [hunk 35]
```diff
@@ -0,0 +1,399 @@
...
+                df_converted["events_start_end_month"] = (
+                    df_converted["events_startMonth.code"] + "-"
+                )
+
+        # Define the principal investigator / contact person, data_sources
+        # and recruitment_sources
+        df_converted = self._convert_list_values(df_converted)
+
+        # Create and fill the principal investigator and contact person columns
+        # TODO: ik krijg dit (nog) niet in een stap.
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/lifecycle.py:226]: After a bit of trial and error, I've managed to create a function that does it in one step.

File: src/molgenis/eucan_connect/model.py (hunk 21 above)
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/model.py:237]: Done

File: src/molgenis/eucan_connect/model.py (hunk 22 above)
Comment (dtroelofsprins) [reply src/molgenis/eucan_connect/model.py:207]: Done

File: tests/conftest.py (hunk 23 above)
Comment (dtroelofsprins) [reply tests/conftest.py:70]: Done

File: tests/test_eucan.py (hunk 24 above)
Comment (dtroelofsprins) [reply tests/test_eucan.py:152]: = Solved

File: tests/test_eucan.py [hunk 36]
```diff
@@ -0,0 +1,248 @@
...
+    #  niet meer
+    # assert mock_get_uploadable_data.mock_calls == [
+    #         mock.call(catalogue, converted_source_data, "events"),
+    #         mock.call(catalogue, converted_source_data, "persons"),
+    #         mock.call(catalogue, converted_source_data, "population"),
+    #         mock.call(catalogue, converted_source_data, "study"),
+    #         ]
+    # TODO: Dit werkt wel, maar kan alleen de laatste checken
+    # mock_get_uploadable_data.assert_called_with(catalogue, converted_source_data,
+    #                                            "study")
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply tests/test_eucan.py:105]: No, finally I found a way that the function within the function still works, but you can also mock it:
You can set the side effect of the mock to be the original method.

File: tests/test_importer.py (hunk 25 above)
Comment (dtroelofsprins) [reply tests/test_importer.py:34]: Solved

File: tests/test_lifecycle.py [hunk 37]
```diff
@@ -0,0 +1,79 @@
...
+    catalogue = Catalogue("LC", "LifeCycle", "lifecycle_url", "LifeCycle")
+    lifecycle.get_lc_cohort_data = MagicMock()
+    lifecycle.get_lc_cohort_data.return_value = lifecycle_data
+    lifecycle._create_df = MagicMock()
+    lifecycle._create_df.return_value = lifecycle_created_df
+    lifecycle._convert_values = MagicMock()
+    lifecycle._convert_values.return_value = lifecycle_created_df
+
+    # ToDo: Also mock and test Lifecycle _convert_list_values, _extract_data and
+    #  _group_column_information?
```
[the remark this answers is not in the document]
Comment (dtroelofsprins) [reply tests/test_lifecycle.py:22]: Done

File: tests/test_lifecycle.py (hunk 26 above)
Comment (dtroelofsprins) [reply tests/test_lifecycle.py:79]: Done

**Comment by tommydeboer** [comment]
[[#8715] Finish EUCAN-Connect PR](https://trello.com/c/Ecze7FKy/427-8715-finish-eucan-connect-pr)

**Review by tommydeboer — APPROVED** [review]

**Pull request merged**